File: ehmp/product/production/hmp-main/src/main/java/gov/va/hmp/jira/JiraAuth.java
package gov.va.hmp.jira;

import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
import org.springframework.web.client.RestTemplate;

import java.io.Serializable;
import java.util.Map;

/**
 * Created with IntelliJ IDEA.
 * User: Jim
 * Date: 1/30/14
 * Time: 10:20 AM
 * To change this template use File | Settings | File Templates.
 */
public class JiraAuth implements Serializable {

    private String username;
    private transient String password;
    private Map<String, Object> project;
    private Map<String, Object> issueType;

    /*
     *  These have to each have their own restTemplate, because they seem to
     */
    private transient RestTemplate restTemplate;

    public RestTemplate getRestTemplate() {
        if(restTemplate==null) {
            restTemplate = new RestTemplate(new HttpComponentsClientHttpRequestFactory());
        }
        return restTemplate;
    }

    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public String getPassword() {
        return password;
    }

    public void setPassword(String password) {
        this.password = password;
    }

    public Map<String, Object> getProject() {
        return project;
    }

    public void setProject(Map<String, Object> project) {
        this.project = project;
    }

    public Map<String, Object> getIssueType() {
        return issueType;
    }

    public void setIssueType(Map<String, Object> issueType) {
        this.issueType = issueType;
    }

    public String getStackTraceFieldId() {

        return stackTraceFieldId;
    }

    public void setStackTraceFieldId(String stackTraceFieldId) {
        this.stackTraceFieldId = stackTraceFieldId;
    }

    private String stackTraceFieldId;
}

File: ehmp/product/production/hmp-main/src/main/js/gov/va/hmp/jira/JiraAuth.js
/**
 * Created with IntelliJ IDEA.
 * User: Jim
 * Date: 2/10/14
 * Time: 5:04 PM
 * To change this template use File | Settings | File Templates.
 */
Ext.define('gov.va.hmp.jira.JiraAuth', {
    singleton: true,
        doWithAuth: function(fn) {
            if(!gov.va.hmp.jira.JiraAuth.sessionAuthenticated) {
                var submit = function(obj) {
                        var wnd = obj.up('window'), frm = wnd.down('form');
                        frm.hide();
                        wnd.setLoading('Authenticating...');
                        Ext.Ajax.request({
                            url: '/jira/authenticate',
                            method: 'POST',
                            params: frm.getValues(),
                            success: function(resp) {
                                var rj = Ext.decode(resp.responseText);
                                if(rj.authenticated=="true") {
                                    gov.va.hmp.jira.JiraAuth.sessionAuthenticated=true;
                                    gov.va.hmp.jira.JiraAuth.username=rj.username;
                                    fn();
                                    wnd.close();
                                } else {
                                    wnd.setLoading(false);
                                    frm.show();
                                    Ext.MessageBox.alert('Authentication Failed','Unable to authenticate; Check your credentials and try again.');
                                }
                            },
                            failure: function(resp) {
                                console.log(resp.responseText);
                                wnd.setLoading(false);
                                frm.show();
                                Ext.MessageBox.alert("Cannot authenticate; Check error log for details");
                            }
                        });
                    },
                    skey = function(obj, keyEvt) {
                        if(keyEvt.getKey()===keyEvt.ENTER) {
                            submit(obj);
                        }
                    },
                    authFrm = Ext.widget('window',{
                        width: 300,
                        height: 190,
                        layout: 'fit',
                        title: 'Jira Login',
                        items: [{
                            xtype: 'form',
                            layout: {
                                type: 'vbox',
                                align: 'stretch'
                            },
                            items: [{
                                xtype: 'component',
                                autoEl: 'p',
                                html: 'Please enter your JIRA credentials to continue.'
                            },{
                                xtype: 'textfield',
                                fieldLabel: 'Username',
                                name: 'username',
                                flex: 0
                            },{
                                xtype: 'textfield',
                                inputType: 'password',
                                fieldLabel: 'Password',
                                name: 'password',
                                flex: 0
                            }]
                        }],
                        fbar: [
                            {
                                ui: 'primary',
                                text: 'Authenticate',
                                handler: submit
                            }
                        ]
                    });
                authFrm.down('form').items.items[0].on('specialkey', skey);
                authFrm.down('form').items.items[1].on('specialkey', skey);
                authFrm.show();
            } else {
                fn();
            }
        }
});
